### PR TITLE
Helper for parsing enums from string

### LINKED ISSUE #258: Create Enum helper for parsing enums
The helper methods should be able to convert the string representation of the name, the numeric value or the value from EnumMember attribute to an equivalent enumerated object.

### PR DESCRIPTION
**Pull request by MartinM85**
Closes #258

Generic method is same as the current method implemented in https://github.com/microsoft/kiota-serialization-json-dotnet/blob/d0454daecb7ec2327f3e55d642fa510cf617e8b5/src/JsonParseNode.cs#L214

### CODE REVIEW

File: src/Helpers/EnumHelpers.cs [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        /// <returns></returns>
+#if NET5_0_OR_GREATER
+        public static T? GetEnumValue<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicFields)] T>(string rawValue) where T : struct, Enum
+#else
+        public static T? GetEnumValue<T>(string rawValue) where T : struct, Enum
+#endif
+        {
+            if(string.IsNullOrEmpty(rawValue)) return null;
+
+            var type = typeof(T);
```
Comment (andrueastman) [inline src/Helpers/EnumHelpers.cs:30]:
```suggestion
```

File: src/Helpers/EnumHelpers.cs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+#else
+        private static string ToEnumRawName<T>(string value) where T : struct, Enum
+#endif
+        {
+            if(TryGetFieldValueName(typeof(T), value, out var val))
+            {
+                value = val;
+            }
+
+            return value;
```
Comment (andrueastman) [inline src/Helpers/EnumHelpers.cs:66]:
```suggestion
            return TryGetFieldValueName(typeof(T), value, out var val) ? val : value;
```

File: src/Helpers/EnumHelpers.cs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        private static ReadOnlySpan<char> ToEnumRawName<T>(ReadOnlySpan<char> span) where T : struct, Enum
+#endif
+        {
+            var value = span.ToString();
+            if(TryGetFieldValueName(typeof(T), value, out var val))
+            {
+                value = val;
+            }
+
+            return value.AsSpan();
```
Comment (andrueastman) [inline src/Helpers/EnumHelpers.cs:81]:
```suggestion
            return TryGetFieldValueName(typeof(T), span.ToString(), out var val) ? val.AsSpan() : span;
```

File: Microsoft.Kiota.Abstractions.Tests/EnumHelperTests.cs [hunk 4]
```diff
@@ -0,0 +1,121 @@
+﻿using Microsoft.Kiota.Abstractions.Helpers;
```
Comment (andrueastman) [inline Microsoft.Kiota.Abstractions.Tests/EnumHelperTests.cs:1]: 👍🏼

**Review by andrueastman — COMMENTED** [review]
Just a few suggestions otherwise looks good from my end.

File: src/Helpers/EnumHelpers.cs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+                }
+#endif
+            }
+            return result;
+
+            
+        }
+
+#if NET5_0_OR_GREATER
+        private static bool TryGetFieldValueName([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicFields)] Type type, string rawValue, out string valueName)
```
[the remark this answers is not in the document]
Comment (MartinM85) [reply src/Helpers/EnumHelpers.cs:142]: But TryGetFieldValueName is called also from ToEnumRawName, so removing the annotation will lead to 

> IL2070	'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicFields' in call to 'System.Type.GetFields()'. The parameter 'type' of method 'Microsoft.Kiota.Abstractions.Helpers.EnumHelpers.TryGetFieldValueName(Type, String, out String)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to

File: src/Helpers/EnumHelpers.cs [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        }
+
+        /// <summary>
+        /// Gets the enum value from the raw value for the given type
+        /// </summary>
+        /// <param name="type">Enum type</param>
+        /// <param name="rawValue">Raw value</param>
+        /// <returns></returns>
+#if NET5_0_OR_GREATER
+        public static object? GetEnumValue([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicFields)] Type? type, string rawValue)
```
[the remark this answers is not in the document]
Comment (andrueastman) [reply src/Helpers/EnumHelpers.cs:79]: Out of curiosity, if we call this from the other `sendPrimitive` method, will this force us to update the annotations of the method as well?

[the remark this answers is not in the document]
Comment (MartinM85) [reply src/Helpers/EnumHelpers.cs:79]: Yes, it will

File: src/Helpers/EnumHelpers.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (baywet) [reply src/Helpers/EnumHelpers.cs:142]: OK thank you i'm very confused by those attributes and the way you're supposed to use them.

File: src/Helpers/EnumHelpers.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (andrueastman) [reply src/Helpers/EnumHelpers.cs:79]: I only ask to know if we should we pre-emptively update the `sendPrimitive` interface as well in preparation? Or would you prefer to have that updated in a separate PR after this is released?

[the remark this answers is not in the document]
Comment (MartinM85) [reply src/Helpers/EnumHelpers.cs:79]: I think it would be good to update both `sendPrimitive` and `sendPrimitiveCollection` interface

**Review by andrueastman — COMMENTED** [review]
This looks good to me.

@MartinM85 Any chance you were looking to update https://github.com/microsoft/kiota-http-dotnet/pull/266 with a local reference of this PR first to validate the changes here before we merge?

**Comment by MartinM85** [comment]
> This looks good to me.
> 
> @MartinM85 Any chance you were looking to update [microsoft/kiota-http-dotnet#266](https://github.com/microsoft/kiota-http-dotnet/pull/266) with a local reference of this PR first to validate the changes here before we merge?

Confirmed...with a local reference of Kiota.Abstractions and enum helper called from SendPrimitiveAsync, the tests added in [microsoft/kiota-http-dotnet#266](https://github.com/microsoft/kiota-http-dotnet/pull/266) passed

**Review by andrueastman — APPROVED** [review]
Looks good from my end. Will merge once @baywet does a final review.

**Review by baywet — APPROVED** [review]

**Pull request merged**